File: C_Kefa_and_Park.cpp
#include <bits/stdc++.h>
using namespace std;

#define int long long

signed main() {
    int n, m, ans = 0;
    cin >> n >> m;
    vector<int> a(n + 1), g[n + 1];
    for (int i = 1; i <= n; ++i) {
        cin >> a[i];
    }
    for (int i = 1; i < n; ++i) {
        int x, y;
        cin >> x >> y;
        g[x].emplace_back(y);
        g[y].emplace_back(x);
    }

    function<void(int, int, int)> dfs = [&](int u, int father, int cnt) {
        if (u ^ 1 and g[u].size() == 1) {
            ++ans;
            return;
        }
        for (int v : g[u]) {
            if (v == father or cnt + a[v] > m) {
                continue;
            }
            if (a[v]) {
                dfs(v, u, cnt + 1);
            } else {
                dfs(v, u, 0);
            }
        }
    };

    dfs(1, -1, a[1]);
    cout << ans << endl;
    return 0;
}

File: C_Kefa_and_Park.py
n, m = map(int, input().split())
a = [-1] + list(map(int, input().split()))
g = [[] for _ in range(n + 1)]
for _ in range(n - 1):
    x, y = map(int, input().split())
    g[x].append(y)
    g[y].append(x)

ans = 0
stack = [[1, -1, a[1]]]
while stack:
    u, father, cnt = stack.pop()
    if u ^ 1 and len(g[u]) == 1:
        ans += 1
    for v in g[u]:
        if v == father or cnt + a[v] > m:
            continue
        if a[v]:
            stack.append([v, u, cnt + 1])
        else:
            stack.append([v, u, 0])

print(ans)
